### ISSUE TITLE
Switch default for `useDefineForClassFields` in ESNext

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#34787]
> Since this targets 3.7.1, I'm going to close this bug. When class fields reaches stage 4, I'll do the final two steps.

_Originally posted by @sandersn in https://github.com/microsoft/TypeScript/issues/27644#issuecomment-542299230_
___

`useDefineForClassFields` should be switched to true when targeting ESNext.

**Comment by sandersn** [comment]
Class fields aren't stage 4 yet. Moving to 3.9.

**Comment by jelbourn** [comment]
This isn't mentioned in the RC notes for 3.9- has it been bumped to a later version?

**Comment by DanielRosenwasser** [comment]
Don't think this was done in 3.9 - @sandersn?

**Comment by sandersn** [comment]
No, class fields still aren't at stage 4. For 4.0, I did make https://github.com/microsoft/TypeScript/pull/37894, which makes suspicious accessor/property override patterns an error everywhere. Together with the included codefix, that finishes off all the planned work for #27644 except this issue.

**Comment by nicolo-ribaudo** [comment]
FYI, we plan to switch the equivalent Babel option to `true` by default in Babel 8.

**Comment by ExE-Boss** [comment]
See also <https://github.com/microsoft/TypeScript/issues/41788>, which will be fixed by this.

**Comment by guilhermesimoes** [comment]
Hey everyone, sorry, I know I shouldn't ask questions on issues (where should I ask them?) but I was wondering one thing.
Is there a way to configure TS so that uninitialised variables use `declare` implicitly?
That way this code
```ts
class SomeClass {
  private prop1 = 'default';
  private prop2?: string
}
```
would emit:
```ts
class SomeClass {
  constructor() {
    this.prop1 = 'default';
  }
}
```
without me having to add `declare` to all my uninitialised properties

**Comment by sandersn** [comment]
@guilhermesimoes this is a perfect question for Stack Overflow since it results in a yes/no answer that is recorded for the future and can be updated if needed.

**Comment by CarterLi** [comment]
#41788 changed default for useDefineForClassFields which broke my code and took me more than 1 hour to find the reason.

It's ok to introduce break changes and it's easy to revert to the old behavior too. The problem is that [the release note said nothing about this](https://devblogs.microsoft.com/typescript/announcing-typescript-4-3/#breaking-changes).

If it's expected, please note it somewhere. Thanks.

**Comment by ExE-Boss** [comment]
This has shipped in **TypeScript 4.3.1‑rc**, so it probably should’ve been kept in the [TypeScript 4.3.1](https://github.com/microsoft/TypeScript/milestone/138) milestone.

**Comment by brainkim** [comment]
Just ran into this as a bug. Was using `target: ESNEXT` and using class properties to define types for classes. I had defined a property on the prototype as an optimization, and when trying to do an upgrade of TypeScript started noticing that the property was overwritten to be `undefined`. Really would appreciate this kind of change appear in the release notes.

Alternatively, if there were a way to define the types of class properties without buying into the whole TC39 class properties spec, that would be fantastic.

**Comment by nicolo-ribaudo** [comment]
> Alternatively, if there were a way to define the types of class properties without buying into the whole TC39 class properties spec, that would be fantastic.

You can write `declare` before the property name.

**Comment by brainkim** [comment]
@nicolo-ribaudo 
I consider `declare` and most modifiers to be a little too noisy, but yeah that’s what I’m doing. Thanks for the tip!

**Comment by ExE-Boss** [comment]
@brainkim
> One potential option would be to allow us to define properties via sibling interface:
> 
> ```ts
> interface Point {
> 	x: number;
> 	y: number;
> }
> 
> class Point {
> 	constructor() {
> 		this.x = 0;
> 		// the absence of this line should cause an error
> 		// this.y = 0;
> 	}
> }
> ```

You can do that already.

It’s used in the `@types/node` package: <https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d8e2ccd84ce78d70e48863e874653e2d8f9b75ae/types/node/events.d.ts#L21-L22>

**Comment by brainkim** [comment]
@ExE-Boss 
But you’re losing the strict property initialization checks right?

**Comment by sandersn** [comment]
Class fields moved to stage 4 in the April 2021 meeting, slated for publication in ES2022.

**Comment by robpalme** [comment]
For reference, this was fixed by PR https://github.com/microsoft/TypeScript/pull/42663 and released in TypeScript 4.3.
